educator/src/Dscp/Educator/Web/Auth.hs

```diff
@@ -0,0 +1,72 @@
+{-# LANGUAGE DataKinds     #-}
+{-# LANGUAGE TypeOperators #-}
+
+-- | Necessary types and implementation for web server authenthication
+module Dscp.Educator.Web.Auth
+       ( Auth'
+       , AuthData (..)
+       , checkAuthData
+       ) where
+
+import Data.Aeson (FromJSON)
+import Data.Time.Clock (UTCTime, diffUTCTime, getCurrentTime)
+import Network.Wai (rawPathInfo)
+import Servant ((:>), HasServer, ServantErr (..), ServerT, err401, hoistServerWithContext, route)
+import Servant.Auth.Server (AuthCheck (..), AuthResult (..), FromJWT)
+import Servant.Auth.Server.Internal.Class (AreAuths (..), runAuths)
+import Servant.Server.Internal.RoutingApplication (DelayedIO, addAuthCheck, delayedFailFatal,
+                                                   withRequest)
+
+---------------------------------------------------------------------------
+-- Data types
+---------------------------------------------------------------------------
+
+-- | Custom authetication API type.
+data Auth' auth res
+
+-- | A type that reperesents the data that the client has sent to authenthicate
+data AuthData = AuthData
+    { adPath :: Text
+    , adTime :: UTCTime
+    }
+    deriving Generic
+
+---------------------------------------------------------------------------
+-- Instances
+---------------------------------------------------------------------------
+
+instance ( HasServer api ctxs
+         , AreAuths '[auth] ctxs (AuthData, res)
+         ) => HasServer (Auth' auth res :> api) ctxs where
+    type ServerT (Auth' auth res :> api) m = res -> ServerT api m
+
+    hoistServerWithContext _ pc nt s = hoistServerWithContext (Proxy :: Proxy api) pc nt . s
+
+    route _ context subserver = route (Proxy :: Proxy api) context (addAuthCheck subserver authCheck)
+      where
+        authMsg :: ByteString
+        authMsg = "Set Authorization header with a proper JWT"
+        authCheck :: DelayedIO res
+        authCheck = withRequest $ \req -> do
+            let authChecks = runAuths (Proxy :: Proxy '[auth]) context
+            authRes <- liftIO $ runAuthCheck (authChecks >>= checkAuthData) req
+            case authRes of
+                (Authenticated (r :: res)) -> return r
+                _ -> delayedFailFatal $ err401 { errHeaders = [("WWW-Authenticate", authMsg)] }
+
+instance FromJSON AuthData
+instance FromJWT AuthData
+
+---------------------------------------------------------------------------
+-- Helpers
+---------------------------------------------------------------------------
+
+checkAuthData :: (AuthData, r) -> AuthCheck r
+checkAuthData (AuthData {..}, r) = do
```

Comment:
This doesn't make any sense. Why do you just return the `r`? This also forces you to write a pretty weird `instance IsAuth EducatorAuth (AuthData, ())`. Can't it be solved trivially with basic tools like typeclasses and adts?

Reply:
But why? What does that achieve?

Reply:
I've already answered that in the first comment.

Reply:
Your first comment basically says "it looks weird" and not much more. That is subjective, because it doesn't to me. I am not convinced that what you're proposing is better, at least not from how I understood it.

I want `HasServer` instance to be generic and work the same for anything that ever wants to use Auth. That way I know for sure that the general part of authentication, `AuthData` is always checked, because it is ensured by this instance. For that to work, we separate the return type of the JWT signature checking into two things: AuthData that is generalized auth and checked by `HasServer` instance and what is the API endpoint handler is actually going to receive which is represented by this tuple.

Besides, this is not much different from how the original student auth worked. Look at `WithStudent` type, it is pretty much the same tuple. That is _exactly the same thing_, but just instead of having `WithStudent AuthData` we now have basically `WithAuth Student` in form of a tuple.

Reply:
>That is subjective

All code standards are. What's better, `instance MyClass (Datatype, (), (), Int)` or `instance MyClass MyDatatype`?

>I want HasServer instance to be generic 

There's no need to want that, but if you _really do_, you can use typeclasses or ADTs. Or, at least, you can write `instance IsAuth EducatorAuth (GenericAuth ())`, which is better than the pair. Pairs are pretty unobvious.

I got it, but is it possible to define a type alias?

Reply:
Sure, yes.